python_modules/libraries/dagster-aws/dagster_aws/athena/resources.py

```diff
@@ -0,0 +1,263 @@
+
+        if retries < 0:
+            raise AthenaTimeout()
+
+        if state != "SUCCEEDED":
+            raise AthenaError(execution["Status"]["StateChangeReason"])
+
+    def _results(self, execution_id):
+        execution = self.client.get_query_execution(QueryExecutionId=execution_id)["QueryExecution"]
+        s3 = boto3.resource("s3")
+        output_location = execution["ResultConfiguration"]["OutputLocation"]
+        bucket = urlparse(output_location).netloc
+        prefix = urlparse(output_location).path.lstrip("/")
+
+        results = []
+        rows = s3.Bucket(bucket).Object(prefix).get()["Body"].read().decode().splitlines()
+        reader = csv.reader(rows)
+        next(reader)  # Skip the CSV's header row
+        for row in reader:
+            results.append(tuple(row))
+
+        return results
+
+
+class FakeAthenaResource(AthenaResource):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.polling_interval = 0
+        self.stubber = Stubber(self.client)
+
+        s3 = boto3.resource("s3")
+        self.bucket = s3.Bucket("fake-athena-results-bucket")
+        self.bucket.create()
+
+    def execute_query(
+        self, query, fetch_results=False, expected_states=None, expected_results=None
+    ):  # pylint: disable=arguments-differ
+        """Fake for execute_query; stubs the expected Athena endpoints, polls against the provided
+        expected query execution states, and returns the provided results as a list of tuples.
+
+        Args:
+            query (str): The query to execute.
+            fetch_results (Optional[bool]): Whether to return the results of executing the query.
+                Defaults to False, in which case the query will be executed without retrieving the
+                results.
+            expected_states (list[str]): The expected query execution states.
+                Defaults to successfully passing through QUEUED, RUNNING, and SUCCEEDED.
+            expected_results ([List[Tuple[Any, ...]]]): The expected results. All non-None items
+                are cast to strings.
+                Defaults to [(1,)].
+
+        Returns:
+            Optional[List[Tuple[Optional[str], ...]]]: The expected_resutls when fetch_resutls is
+                set. Otherwise, return None. All items in the tuple are represented as strings except
+                for empty columns which are represented as None.
+        """
+        if not expected_states:
+            expected_states = ["QUEUED", "RUNNING", "SUCCEEDED"]
+        if not expected_results:
+            expected_results = [("1",)]
+
+        self.stubber.activate()
+
+        execution_id = str(uuid.uuid4())
+        self._stub_start_query_execution(execution_id, query)
+        self._stub_get_query_execution(execution_id, expected_states)
+        if expected_states[-1] == "SUCCEEDED" and fetch_results:
+            self._fake_results(execution_id, expected_results)
```

Comment:
L131-134: nice implementation! seems like an area where moto doesnt provide direct support for. im a bit worried about api divergence down the road, but would assume its reasonably stable

Reply:
I suspect that even if moto fully mocks the API, they won't support state transitions and will continue to recommend that we fiddle with the [moto backend to do that kind of stuff](https://github.com/spulec/moto/issues/1524#issuecomment-652801523).

Localstack's pro version provides a containerized Athena service (I think they more or less run Presto and front it with a small server that mocks the Athena API) but it's a paid product so I doubt it could even be distributed with Dagster and it doesn't look like Dagster makes much use of containerized services for tests at the moment.

But yeah, I'm comfortable assuming this API is relatively stable at this point. Using the botocore Stubber should at least guard against us making invalid calls/responses in our mocks.